### ISSUE TITLE
Unknown Error: 401, MSG: invalid_auth On open Socket  request

### ISSUE THREAD

**Issue by Nazar65** [emacs-slack/emacs-slack#584]
So it started to be few days ago after the system update. 

Emacs Version is 28.2
 i'm using guix gnu/linux.

So the slack-start is working ok, but then when it  do attempt to sync data tthe following error is appearing:
```

[2023-09-20 11:53:25] [warn] Reconnecting... [1/360] - workspace
[2023-09-20 11:53:25] [debug] Authorization Finished - workspace
[2023-09-20 11:53:25] [debug] Opening websocket connection. NOWAIT: nil, URL: wss://wss-primary.slack.com/websocket/LEGACY_BOT:91eb05033d26eac6d622d5acf9346a92 - workspace
[2023-09-20 11:53:25] [debug] Called `websocket-open' URL: wss://wss-primary.slack.com/websocket/LEGACY_BOT:91eb05033d26eac6d622d5acf9346a92 - workspace
[2023-09-20 11:53:25] [trace] REQUEST FINISHED. URL: "https://slack.com/api/rtm.connect", PARAMS: (("mpim_aware" . "1") ("presence_sub" . "true")), DATA: (:ok t :url "wss://wss-primary.slack.com/websocket/LEGACY_BOT:91eb05033d26eac6d622d5acf9346a92" :team (:id "T036RJZEB" :name "name" :domain "domain") :self (:id "XXXXXXX" :name "name")) - workspace
[2023-09-20 11:53:26] [debug] WebSocket on-open - workspace
[2023-09-20 11:53:26] [debug] Unknown Error: 401, MSG: invalid_auth - workspace
[2023-09-20 11:53:26] [debug] Websocket on-close: STATE: closed - workspace
```

Which is "Unknown Error: 401, MSG: invalid_auth - workspace" 401 - Unathorized

**Comment by unhammer** [comment]
I got this, tried adding `; d-s=`  as described in https://github.com/yuya373/emacs-slack/issues/566#issuecomment-1208866953 which let me connect (but still seeing https://github.com/yuya373/emacs-slack/issues/555 ). I'm guessing something changed at Slack.

**Comment by Nazar65** [comment]
@unhammer Just discovered that sockets are okay in browser in slack account, with all those cookies, so probably need to figure-out which one is required

**Comment by Nazar65** [comment]
Okay, i have resolved an issue, so it is now required to send d cookie also with the websocket requests, so it simply need to set cookie like this:
`(url-cookie-store "d" "xoxd-xxxxx" nil ".slack.com" "/" t)`

**Comment by Nazar65** [comment]
@unhammer That should fix both errors with invalid_auth and callback error

**Comment by unhammer** [comment]
What `request-backend` are you using? I think for me, slack puts my cookies in ~/.emacs.d/request/curl-cookie-jar

(Does it solve #555 for you too?)

**Comment by Nazar65** [comment]
@unhammer Yes that is cookie for the curl requests, but slack changed something and now it requires to put d cookie to the websocket requests. And yes it resolves for me #555 also. which was "[error] request--callback: peculiar error: 429"

**Comment by jamesaimonetti** [comment]
Confirming for me as well; thanks for sharing!

**Comment by seblemaguer** [comment]
it works for me for one team, but I don't see how to use it for multiple teams. Did anyone tried with multiple teams to be used in parallel? Thanks a lot :)

**Comment by Nazar65** [comment]
@seblemaguer I;m using only one team, for multiple probably code changes required, to change cookie per each request. also here is more information https://github.com/yuya373/emacs-slack/issues/555#issuecomment-1729478866

**Comment by seblemaguer** [comment]
that makes sense. Thanks for the reply

**Comment by unhammer** [comment]
I've been running this for some days now (setting d and d-s with url-cookie-store) and am switching between teams, but I've only set d/d-s with values from the one team. I haven't had more peculiar error messages, so it seems like they can be reused? But I'm guessing the thing to do is to either change it when switching teams, or just set it per domain (if they differ per team).

**Comment by jumper047** [comment]
@seblemaguer can you test my fix? It should work for multiple accounts, though at this time I have only one and can't test it properly.

**Comment by seblemaguer** [comment]
@jumper047 it looks like it works for me :). It seems also that using one cookie for all the teams work as well. (so only the token differ per team now?)

**Comment by jumper047** [comment]
@seblemaguer apparently, but we are unable to say if it is so intentionally or not. If second then slack may fox it in some random moment in future

**Comment by jumper047** [comment]
Strange - seems like my fix is working very unstable for me, sometimes works and sometimes not. Not sure why. @unhammer , @seblemaguer have you experienced something like that? Also, is it correct that every slack team has their own subdomain like `someteam.slack.com`?

**Comment by seblemaguer** [comment]
@jumper047 your patch works perfectly for me :) I am using the same cookie (xox-d and d-s values) for all the teams. I don't quite understand what changed on slack's side, but maybe give it a try by picking one cookie value working for a team and apply it to the others. 

Hope it helps

**Comment by unhammer** [comment]
Yeah there was a day where it didn't work, and then it started working again; I didn't investigate. It works now ¯\_(ツ)_/¯

**Comment by isamert** [comment]
> Strange - seems like my fix is working very unstable for me,
> sometimes works and sometimes not.

@jumper047 I remember (in some comment somewhere, don't remember where
exactly) that you were having re-connection problems with the
websocket, right?  I also started seeing them and as you (or someone
else?) stated before, restarting emacs fixes the reconnection problem
but that's not very feasible solution. I tried to dig deeper and found
out setting

```emacs-lisp
;; These are used by websocket.el, not request.el related
(setq url-cookie-storage '())
(setq url-cookie-secure-storage '())
```

before the re-connection attempts fixes (also *maybe* clearing the
`url-cookie-file`) the problem, at least reconnection happens
successfully. I did not dig deeper which cookie is causing the problem
but I will investigate it further. In meantime adding the code above
to reconnection logic (in `slack-websocket.el`) might fix it but be
careful as it may break other things using these cookies.
